File: AgeOfArenaUnity/Assets/Scripts/CivilizationDefs.cs
/// <summary>
/// Civilization identity and per-civ bonuses. Each team picks a civ at game start
/// (or uses Default/None). Bonuses are data-only; systems read them via
/// <see cref="CivilizationDefs.Get"/> and apply multipliers in their own tick logic.
/// This keeps the civ layer additive — no civ needs bespoke system code.
///
/// CIVX: this file is the single canonical source of every civ's ID and display
/// string. HUD/selection/wiki all read <see cref="CivBonus.display"/> from here —
/// never hard-code a civ name elsewhere. docs/wiki/06-civilizations.md mirrors these.
/// </summary>
public enum Civilization
{
    None,
    Franks, Britons, Mongols, Japanese, Byzantines,   // original 5
    Aztecs, Teutons, Persians, Vikings, Saracens,      // M9/CIVC expansion (10 playable total)
}

/// <summary>
/// CIVM: a civilization's TEAM bonus — a benefit that (once alliances land in M11)
/// is shared with every allied team. Exposed through
/// <see cref="GameManager.TeamSharedBonus"/> and consumed by gameplay systems.
/// </summary>
public struct TeamBonus
{
    public float gatherFoodBonus;   // additive fraction to team food gather (0.05 = +5%)
}

public struct CivBonus
{
    public Civilization civ;
    public string display;

    // Economy
    public float gatherFoodMult;   // Franks: +20% food gather
    public float gatherWoodMult;
    public float gatherGoldMult;

    // Military
    public float cavalryHpMult;    // Franks: cavalry +20% HP
    public float archerRangeBonus; // Britons: archer +1 range
    public float cavalrySpeedMult; // Mongols: cavalry +25% speed
    public float infantryAttackMult;// Japanese: infantry +10% atk
    public float buildingHpMult;   // Byzantines: buildings +10% HP
    public float healRateMult;     // Byzantines: Medic/Monk +50% heal

    // Economy QoL
    public float farmDecayMult;    // 1.0 = normal; Franks 0.5 = half decay

    // ── M9/CIVD: new identity fields (each read by ≥1 system) ────────────────
    public float archerAttackMult;  // Vikings/Saracens: archer-class +atk (UnitEntity.AttackDamage)
    public float unitTrainTimeMult; // Mongols/Aztecs: faster training, <1 (TrainingQueue.Enqueue)

    // ── M9/CIVM: team (shared) bonus ─────────────────────────────────────────
    public TeamBonus teamBonus;
}

public static class CivilizationDefs
{
    // Helper: build a fully-populated row with sensible neutral defaults, then override.
    static CivBonus Row(Civilization c, string name,
        float food = 1f, float wood = 1f, float gold = 1f,
        float cavHp = 1f, float archRange = 0f, float cavSpd = 1f,
        float infAtk = 1f, float bldHp = 1f, float heal = 1f, float farmDecay = 1f,
        float archAtk = 1f, float trainTime = 1f, float teamFood = 0f) => new()
    {
        civ = c, display = name,
        gatherFoodMult = food, gatherWoodMult = wood, gatherGoldMult = gold,
        cavalryHpMult = cavHp, archerRangeBonus = archRange, cavalrySpeedMult = cavSpd,
        infantryAttackMult = infAtk, buildingHpMult = bldHp, healRateMult = heal,
        farmDecayMult = farmDecay, archerAttackMult = archAtk, unitTrainTimeMult = trainTime,
        teamBonus = new TeamBonus { gatherFoodBonus = teamFood },
    };

    static readonly CivBonus[] Table =
    {
        Row(Civilization.None,       "Yok"),
        Row(Civilization.Franks,     "Franklar",      food: 1.2f, cavHp: 1.2f, farmDecay: 0.5f),
        Row(Civilization.Britons,    "Britanyalılar", wood: 1.15f, archRange: 1f),
        Row(Civilization.Mongols,    "Moğollar",      gold: 1.1f, cavSpd: 1.25f, trainTime: 0.9f),
        Row(Civilization.Japanese,   "Japonlar",      wood: 1.1f, infAtk: 1.1f),
        Row(Civilization.Byzantines, "Bizanslılar",   bldHp: 1.1f, heal: 1.5f),
        // ── M9/CIVC expansion ──
        Row(Civilization.Aztecs,     "Aztekler",      food: 1.15f, heal: 1.2f, trainTime: 0.9f, teamFood: 0.05f),
        Row(Civilization.Teutons,    "Tötonlar",      infAtk: 1.05f, bldHp: 1.15f),
        Row(Civilization.Persians,   "Persler",       food: 1.1f, cavHp: 1.1f),
        Row(Civilization.Vikings,    "Vikingler",     archAtk: 1.1f, wood: 1.1f),
        Row(Civilization.Saracens,   "Saracenler",    gold: 1.15f, archAtk: 1.1f),
    };

    public static CivBonus Get(Civilization c)
    {
        for (int i = 0; i < Table.Length; i++)
            if (Table[i].civ == c) return Table[i];
        return Table[0];
    }

    /// <summary>All civilizations a player may pick (excludes <see cref="Civilization.None"/>).</summary>
    public static System.Collections.Generic.IEnumerable<CivBonus> Playable()
    {
        for (int i = 0; i < Table.Length; i++)
            if (Table[i].civ != Civilization.None) yield return Table[i];
    }
}

File: web/src/core/CivilizationDefs.ts
/**
 * CivilizationDefs.cs port — civ identities and per-civ bonuses.
 * Bonuses are data-only; systems read them and apply multipliers in their own tick.
 */

export const enum Civilization {
  None,
  Franks, Britons, Mongols, Japanese, Byzantines,
  Aztecs, Teutons, Persians, Vikings, Saracens,
  Celts, Chinese, Goths, Turks,
}

export interface CivBonus {
  civ:                  Civilization;
  display:              string;
  flag:                 string; // emoji flag for map-select UI

  // Economy
  gatherFoodMult:       number; // Franks: 1.2
  gatherWoodMult:       number;
  gatherGoldMult:       number;
  farmDecayMult:        number; // 1.0 = normal; Franks 0.5 = half decay

  // Military
  cavalryHpMult:        number; // Franks: 1.2
  archerRangeBonus:     number; // Britons: +1
  cavalrySpeedMult:     number; // Mongols: 1.25
  infantryAttackMult:   number; // Japanese: 1.1
  buildingHpMult:       number; // Byzantines: 1.1
  healRateMult:         number; // Byzantines: 1.5

  // New in M9
  archerAttackMult:     number; // Vikings/Saracens
  unitTrainTimeMult:    number; // Mongols/Aztecs: 0.8 = 20% faster

  // Shared team bonus
  teamGatherFoodBonus:  number; // additive fraction (0.05 = +5%)
}

function make(
  civ: Civilization,
  display: string,
  flag: string,
  overrides: Partial<CivBonus> = {},
): CivBonus {
  return {
    civ, display, flag,
    gatherFoodMult: 1, gatherWoodMult: 1, gatherGoldMult: 1, farmDecayMult: 1,
    cavalryHpMult: 1, archerRangeBonus: 0, cavalrySpeedMult: 1,
    infantryAttackMult: 1, buildingHpMult: 1, healRateMult: 1,
    archerAttackMult: 1, unitTrainTimeMult: 1, teamGatherFoodBonus: 0,
    ...overrides,
  };
}

export const CIVILIZATION_DEFS: Record<Civilization, CivBonus> = {
  [Civilization.None]:       make(Civilization.None,       "Belirsiz",   "⬜"),
  [Civilization.Franks]:     make(Civilization.Franks,     "Franklar",   "🇫🇷", { gatherFoodMult: 1.2, cavalryHpMult: 1.2, farmDecayMult: 0.5 }),
  [Civilization.Britons]:    make(Civilization.Britons,    "Britonyalılar","🇬🇧", { archerRangeBonus: 1 }),
  [Civilization.Mongols]:    make(Civilization.Mongols,    "Moğollar",   "🇲🇳", { cavalrySpeedMult: 1.25, unitTrainTimeMult: 0.8 }),
  [Civilization.Japanese]:   make(Civilization.Japanese,   "Japonlar",   "🇯🇵", { infantryAttackMult: 1.1 }),
  [Civilization.Byzantines]: make(Civilization.Byzantines, "Bizanslılar","🇬🇷", { buildingHpMult: 1.1, healRateMult: 1.5 }),
  [Civilization.Aztecs]:     make(Civilization.Aztecs,     "Aztekler",   "🇲🇽", { unitTrainTimeMult: 0.85, teamGatherFoodBonus: 0.05 }),
  [Civilization.Teutons]:    make(Civilization.Teutons,    "Tötonlar",   "🇩🇪", { buildingHpMult: 1.15, cavalryHpMult: 1.1 }),
  [Civilization.Persians]:   make(Civilization.Persians,   "Persler",    "🇮🇷", { gatherFoodMult: 1.05, gatherGoldMult: 1.05 }),
  [Civilization.Vikings]:    make(Civilization.Vikings,    "Vikingler",  "🇳🇴", { archerAttackMult: 1.1 }),
  [Civilization.Saracens]:   make(Civilization.Saracens,   "Sarazenler", "⚔️", { archerAttackMult: 1.05, gatherGoldMult: 1.1 }),
  [Civilization.Celts]:      make(Civilization.Celts,      "Keltler",    "🏴󠁧󠁢󠁳󠁣󠁴󠁿", { infantryAttackMult: 1.15, gatherWoodMult: 1.15 }),
  [Civilization.Chinese]:    make(Civilization.Chinese,    "Çinliler",   "🇨🇳", { gatherFoodMult: 1.1, teamGatherFoodBonus: 0.05 }),
  [Civilization.Goths]:      make(Civilization.Goths,      "Gotlar",     "⚔️", { infantryAttackMult: 1.2, unitTrainTimeMult: 0.75 }),
  [Civilization.Turks]:      make(Civilization.Turks,      "Türkler",    "🇹🇷", { gatherGoldMult: 1.2, unitTrainTimeMult: 0.9 }),
};

/** Civs available in the selection screen (original 5 + M9 expansion). */
export const PLAYABLE_CIVS: Civilization[] = [
  Civilization.Franks, Civilization.Britons, Civilization.Mongols,
  Civilization.Japanese, Civilization.Byzantines,
  Civilization.Aztecs, Civilization.Teutons, Civilization.Persians,
  Civilization.Vikings, Civilization.Saracens,
  Civilization.Celts, Civilization.Chinese, Civilization.Goths, Civilization.Turks,
];
